### PR TITLE
ASoC: SOF: support topology components on slave cores

### PR DESCRIPTION
**Pull request by lyakh**
This is still WiP because it still doesn't work perfectly. It goes together with https://github.com/thesofproject/sof/pull/2513 . Currently slave core support is broken in SOF, but with these two PRs I am able to run e.g. audio playback on an HDMI output on a slave core. However, in less than 50% of the cases I'm still getting DMA underruns when starting the playback on the link DMA channel. In such cases data is first copied from the input buffer over host DMA, then a check is performed in `hda_dma_data_size()` called from `dai_copy()` to get data size, and that call detects the overrun bit set in the DGCS register. Investigating

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -1723,6 +1801,21 @@ static int sof_widget_load_pga(struct snd_soc_component *scomp, int index,
...
+		dev_err(scomp->dev, "error: parse volume.comp tokens failed %d\n",
+			ret);
+		goto err;
+	}
+
+	ret = sof_core_enable(sdev, volume->comp.core);
+	if (ret < 0) {
+		dev_err(scomp->dev, "error: enable core: %d\n", ret);
+		goto err;
+	}
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: @lyakh I am wondering if it makes sense to maybe move the core_tokens parsing to sof_widget_ready() and pass the core_id as a parameter to all the widget load functions? Then, we could avoid adding it so many places

Comment (lyakh) [reply sound/soc/sof/topology.c]: @ranj063 this violates the current approach of parsing all tokens within each load function separately, but I agree, this would save a lot of code and consolidate the processing in one place, will change, thanks!

File: include/uapi/sound/sof/tokens.h [hunk 2]
```diff
@@ -73,6 +73,7 @@
 /* Token retired with ABI 3.2, do not use for new capabilities
```
Comment (kv2019i) [inline include/uapi/sound/sof/tokens.h:73]: commit: "or individual topology components" ?

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -1233,6 +1240,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+	struct sof_ipc_pm_core_config pm_core_config = {
+		.hdr = {
+			.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CORE_ENABLE,
+			.size = sizeof(pm_core_config),
+		},
+		.enable_mask = sdev->enabled_cores_mask | 1 << core,
+	};
+	int ret;
+
+	if (sdev->enabled_cores_mask & 1 << core)
```
Comment (kv2019i) [inline sound/soc/sof/topology.c]: @lyakh BIT(core) ..?

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -1233,6 +1240,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+		return 0;
+
+	/* power up the core */
+	ret = snd_sof_dsp_core_power_up(sdev, 1 << core);
```
Comment (kv2019i) [inline sound/soc/sof/topology.c]: ditto

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -1475,50 +1539,21 @@ int sof_load_pipeline_ipc(struct device *dev,
...
-		dev_err(dev, "error: core enable ipc failure\n");
+		dev_err(dev, "error: load pipeline ipc failure\n");
 
 	return ret;
 }
 
-static int sof_widget_load_pipeline(struct snd_soc_component *scomp,
-				    int index, struct snd_sof_widget *swidget,
+static int sof_widget_load_pipeline(struct snd_soc_component *scomp, int index,
+				    struct snd_sof_widget *swidget,
```
Comment (kv2019i) [inline sound/soc/sof/topology.c:1630]: Unrelated change?

File: sound/soc/sof/topology.c [hunk 6]
```diff
@@ -2230,6 +2275,25 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
...
+	ret = sof_parse_tokens(scomp, &comp, core_tokens,
+			       ARRAY_SIZE(core_tokens), tw->priv.array,
+			       le32_to_cpu(tw->priv.size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: parse dai.comp tokens failed %d\n",
+			ret);
+		return ret;
+	}
+
+	if ((int)comp.core < 0) {
```
Comment (kv2019i) [inline sound/soc/sof/topology.c]: Hmm, this looks iffy. If comp.core is not signed, I'd have a separate more explicit safe conversion from the unsigned field to the integer value (i.e. must be under certain max value).

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:1630]: well, there are two of those changes here, they just align function arguments with other similar functions, which this patch does modify. This wouldn't merit a separate patch even with my ability to submit patches to remove superfluous parentheses or variable initialisations ;-) But sure, if undesired, I can drop these 2 changes.

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: haha, we're much better at seeing errors in others' work - I just made the same comment in someone else's patches :-)

File: include/uapi/sound/sof/tokens.h (hunk 2 above)
Comment (lyakh) [reply include/uapi/sound/sof/tokens.h:73]: well, this hasn't been tested yet. I presume it should work too, for that to work only changes in topology files *should* be needed, but I usually avoid claiming something, that hasn't been verified to work...

File: sound/soc/sof/topology.c (hunk 6 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: it actually just checks, if it has been modified from the initialisation value, I can in fact even just check `== -ENOENT`. Well, if `comp.core` becomes 16 bits at some point, then this will fail, yes... Is this what you're concerned about? I think the safest would be just to initialise it to 0 then and power it up unconditionally.

File: sound/soc/sof/topology.c [hunk 7]
```diff
@@ -1233,6 +1240,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+
+	/* Now notify DSP that the core has been powered up */
+	ret = sof_ipc_tx_message(sdev->ipc, pm_core_config.hdr.cmd,
+				 &pm_core_config, sizeof(pm_core_config),
+				 &pm_core_config, sizeof(pm_core_config));
+	if (ret < 0)
+		dev_err(sdev->dev, "error: core %d enable ipc failure %d\n",
+			core, ret);
+
+	return ret;
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1322]: @lyakh can we just return sof_ipc_tx_message() here? There will be an error printed in that function anyway no?

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -2230,6 +2275,25 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		strnlen(tw->sname, SNDRV_CTL_ELEM_ID_NAME_MAXLEN) > 0
 			? tw->sname : "none");
 
+	ret = sof_parse_tokens(scomp, &comp, core_tokens,
+			       ARRAY_SIZE(core_tokens), tw->priv.array,
+			       le32_to_cpu(tw->priv.size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: parse dai.comp tokens failed %d\n",
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: copy/paste error here. parse core tokens error?

File: sound/soc/sof/topology.c [hunk 9]
```diff
@@ -1233,6 +1240,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		pipeline = swidget->private;
+	} else {
+		pipeline = sof_pipeline_find(sdev, swidget->pipeline_id);
+		if (!pipeline)
+			return -ENOENT;
+	}
+	core = pipeline->core;
+
+	return sof_core_enable(sdev, core);
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: @lyakh i could be wrong but wouldnt this be a duplicate call to sof_core_enable(). We already have a sof_core_enable() in sof_widget_ready()?

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:1322]: @ranj063 well not in all cases. Some failures don't print an error message. Those are really exotic ones like invalid pointers etc. Some calls to it print an error message, some don't. I'd keep it for now, but can remove it too if it's really undesired.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: fixed

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: @ranj063 well, I don't know. What if in a pipeline all components are placed on core 0 and the scheduler widget has 0 for component core and 1 for scheduling core? This is certainly a weird example. Besides, that function is extremely light-weight if the core is already enabled.

File: include/uapi/sound/sof/tokens.h (hunk 2 above)
Comment (kv2019i) [reply include/uapi/sound/sof/tokens.h:73]: @lyakh I think you misunderstood a bit. To me "separate topology components" and "individual topology components" means the same thing. I was just proposing the latter as a better way to express it. But now I'm not anymore what you mean by separate here? I thought the intent is to run the full pipeline, or a subset of the pipeline, right?

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c:1630]: @lyakh Ack, you can keep it. Just confirming it's intentional.

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c]: @lyakh @ranj063 If there's a theoretical chance, let's keep it. It's not like we have rules laid out anywhere currently, so this could come and haunt us later.

File: include/uapi/sound/sof/tokens.h (hunk 2 above)
Comment (lyakh) [reply include/uapi/sound/sof/tokens.h:73]: ah, ok, yes, "individual" sounds better. Yes, that's exactly what I meant. In my comment above I just clarified, that so far only pipelines, running on slave cores, have been tested, not individual components. But the same code in this patch should also support individual components / pipeline subsets, this just hasn't been tested yet.

**Comment by lyakh** [comment]
> Is this still [WiP] @lyakh ? Maybe drop that from summary to lure in more review attention.
> 
> It starts to look good to me and tests are passing, so should we just go ahead and push for merge. I did have one question about the commit message, so if you can check that.

@kv2019i updated.

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/topology.c]: @lyakh @kv2019i  I would argue the other way around. In the past, we have seen FE hangs because we tried to power up a core that was already up. And because you call snd_sof_dsp_power_up() without checking if it is already up, I am afraid we could run into similar issues again. So even if there is a slight chance that this could happen, we should prevent it

Comment (lyakh) [reply sound/soc/sof/topology.c]: @ranj063 there is a *VERY* explicit check for that:
```
	if (sdev->enabled_cores_mask & BIT(core))
		return 0;
```
or do you see any situations where this might not trigger?

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c]: @lyakh Indeed, happens to me all the time :)

File: include/uapi/sound/sof/tokens.h (hunk 2 above)
Comment (kv2019i) [reply include/uapi/sound/sof/tokens.h:73]: Ack, thanks @lyakh

File: sound/soc/sof/topology.c [hunk 10]
```diff
@@ -1233,6 +1243,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+		},
+		.enable_mask = sdev->enabled_cores_mask | BIT(core),
+	};
+	int ret;
+
+	if (sdev->enabled_cores_mask & BIT(core))
+		return 0;
+
+	/* power up the core */
+	ret = snd_sof_dsp_core_power_up(sdev, BIT(core));
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1304]: this doesn't scale to e.g. TGL, where only core0 can be powered-up by the driver. Other cores will have to be powered with an IPC sent to core0.

Comment (ranj063) [reply sound/soc/sof/topology.c:1304]: @plbossart this piece of code exists even in the code today but out of curiosity, what happens when the driver tries to power up one of the salve cores this way?

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/topology.c]: @lyakh sorry I completely missed this check. I only looked in the core_power_up() op to see if the check was there :)

Comment (ranj063) [reply sound/soc/sof/topology.c]: > @ranj063 well, I don't know. What if in a pipeline all components are placed on core 0 and the scheduler widget has 0 for component core and 1 for scheduling core? This is certainly a weird example. Besides, that function is extremely light-weight if the core is already enabled.

@lyakh this would be very weird. Besides for my education, what is the real difference between a comp core and a schedule core? Is a schedule core the core that the pipeline is supposed to run on? Is it possible for the comp core for the components in the pipeline to be different from the pipeline sched core at all?

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:1304]: no idea. timeout or nothing, either way it's not going to be successful.

Comment (lyakh) [reply sound/soc/sof/topology.c:1304]: @plbossart @RanderWang AFAIK the TGL milti-core support is still in the works

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: @ranj063 As far as I understand originally only a pipeline-level multi-core support was implemented, i.e. you could only specify cores per pipeline. That's what the scheduling core is for. Recently a capability has been added to run individual components on different cores. I still don't know what use-cases this should be helpful for. I asked for examples, but I don't think I've got any back. So, now in theory it should be possible to run a pipeline on slave core 1 and a certain component from it on core 2. That means we now have to specify cores per component for each component.
As for this specific examples - all components are marked for cores other than the scheduling core - as I said, I don't know if this is meaningful at all. In theory you can of course imagine such cases: you can have your timer interrupt for the pipeline fire on one core and all DMA copying done on other cores. On our current hardware I'm not aware of any good reasons to do that.

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:1304]: @lyakh  Now the default DSP frequency is in low mode, how about to set it in high mode ? I just got this message from @keyonjie  and don't know how to change it.

Comment (lyakh) [reply sound/soc/sof/topology.c:1304]: @RanderWang sorry, I'm not sure where that frequency is configured either and how is it related to this PR?

Comment (paulstelian97) [reply sound/soc/sof/topology.c:1304]: @RanderWang You need to check the firmware Kconfig and disable the config for low frequency clock, and rebuild the firmware.

**Comment by kv2019i** [comment]
@lyakh Btw, you have not request re-reviews from people. Please do that. (and yes, we should somehow make that automatic as everybody seems to be not doing this)

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c:1304]: Full TGL support is beyond this patch (right @plbossart), but do we need some checks to this patch so that it does not try to do illegal actions on TGL (only core0 should be powered by driver). @lyakh ?

Comment (plbossart) [reply sound/soc/sof/topology.c:1304]: I have a conceptual issue with a patch to support multi-core that doesn't address a known architecture. That essentially means we are going to undo-redo what we could have done right in the first place with more thinking.

It could very well be that the _power_up() function hides the actual enablement and dependency on IPC, but I'd like at least to see that this patch is compatible with directions for TGL. Ignoring TGL is not an option.

Comment (lyakh) [reply sound/soc/sof/topology.c:1304]: @plbossart my understanding is, that multi-core on TGL isn't yet supported by the firmware, see e.g. https://github.com/thesofproject/sof/pull/2516 and as @ranj063 noted, this patch doesn't change the way SOF currently works or doesn't work om TGL specifically. As long as the topology doesn't specify cores other than 0, nothing will happen.

Comment (plbossart) [reply sound/soc/sof/topology.c:1304]: @lyakh my understanding is that TGL multi-core support was a development that was supposed to happen very soon. @lbetlej @mmaka1 am I missing something here?

**Comment by lyakh** [comment]
> @lyakh Btw, you have not request re-reviews from people. Please do that. (and yes, we should somehow make that automatic as everybody seems to be not doing this)

@kv2019i indeed, as soon as a new version is submitted, I'd expect all reviews for the previous version to be canceled and (if that's how this process works) reviews should be re-requested. With email reviews this is obvious. You have a previous version, you have reviews for it, you have a new version, you can compare the two versions check them against the previous reviews to see how they were addressed. All this is impossible on GH.

File: sound/soc/sof/sof-priv.h [hunk 11]
```diff
@@ -54,6 +54,9 @@ extern int sof_core_debug;
 	(IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE) || \
 	 IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST))
 
+/* So far the master core on all DSPs has ID 0 */
+#define SOF_DSP_MASTER_CORE 0
```
Comment (abonislawski) [inline sound/soc/sof/sof-priv.h:58]: its not possible to get this from platform files or something?

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (abonislawski) [reply sound/soc/sof/topology.c:1304]: @plbossart  yes, its on ready-to-merge stage so please review kernel support for TGL.
it will not work If in TGL  snd_sof_dsp_ops .core_power_up = hda_dsp_enable_core

**Comment by kv2019i** [comment]
> @kv2019i indeed, as soon as a new version is submitted, I'd expect all reviews for the previous version to be canceled and (if that's how this process works) reviews should be re-requested. With email reviews this is obvious. You have a previous version, you have reviews for it, you have a new version, you can compare the two versions check them against the previous reviews to see how they were addressed. All this is impossible on GH.

Or we don't know how to use GH properly. :) There's one slight difference to email lists, and that is that pushing a new revision of patches also triggers CI. I can see some PR updates are pushed to see if a previously reported CI issue is fixed, or to let validation test a new revision, and not necessarily a patch that is ready for review (e.g. author knows about some style issues).

In Gerrit (which is very similar to GH), there was the concept of author+1, which was really good for this. To signal others the patch is ready for reviews, author needed to give his own +1 first. In GH, there is the "Re-request reviews" function (and also "Draft" status and tags), but I liked the Gerrit option better. I could quickly see from the issue whether author claims the most recent revision of the patchset to be ready or not. In GH, I'm trying to look at the the last comments from author and whether he requested re-reviews. I now started to simply look at all PRs that have been updated, but the record from this week is that >50% of the PRs I look every day, are updated but, for a reason or another, are not ready for review, so I'll skip them. I'll now use this method for some time and see how it works. At least I don't miss any PR that is waiting for action.

Of course, in all systems, it's the patch authors responsibility to make sure patch moves forward. If CI fails, reviewers don't provide feedback, it's always the patch author that needs to resend and/or ping the maintainers to get the patch moving.

FYI @plbossart @ranj063

**Comment by lyakh** [comment]
> I now started to simply look at all PRs that have been updated, but the record from this week is that >50% of the PRs I look every day, are updated but, for a reason or another, are not ready for review, so I'll skip them. I'll now use this method for some time and see how it works. At least I don't miss any PR that is waiting for action.

@kv2019i that's what I'm doing too: going to the PR list, sorting them by "last updated" and so you have all the PRs marked that have had some activity since you last visited them. And it's not only good for updated patches but also for ongoing discussions.

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:1304]: @abonislawski sorry, has it been merged by now and I've forgotten or have you got a link to that PR?

File: sound/soc/sof/sof-priv.h (hunk 11 above)
Comment (lyakh) [reply sound/soc/sof/sof-priv.h:58]: @abonislawski I don't see a good place for this ATM. On Intel / HDA platforms we could add this to `struct sof_intel_dsp_desc` but it isn't used by other arches (iMX8). For now this is a place-holder which is at least marking the potential need for making this more flexible instead of assuming the value implicitly.

**Comment by lyakh** [comment]
> The logic and code look good for APL..ICL, but will not work for TGL/EKL. we need a different IPC-based mechanism to turn cores on/off.

@plbossart I still don't see a PR, adding multi-core support to TGL. If it hasn't been submitted yet, I think it would be better to merge this one and then add core ID support to the TGL PR. I agree that this PR is preferable to #1981

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (abonislawski) [reply sound/soc/sof/topology.c:1304]: @lyakh not merged yet
https://github.com/thesofproject/sof/pull/2516

File: sound/soc/sof/sof-audio.c [hunk 12]
```diff
@@ -142,6 +142,22 @@ static int sof_restore_kcontrols(struct device *dev)
...
+		if (swidget->id == snd_soc_dapm_scheduler) {
+			const struct sof_ipc_pipe_new *pipeline =
+				swidget->private;
+			if (pipeline->pipeline_id == pipeline_id)
+				return pipeline;
+		}
+
+	return NULL;
+}
+
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:160]: wasn't this also in the virtio helper functions? If yes we should move it as a separate patch.

File: sound/soc/sof/sof-audio.c [hunk 13]
```diff
@@ -161,6 +177,8 @@ int sof_restore_pipelines(struct device *dev)
 		if (!swidget->private)
 			continue;
 
+		sof_pipeline_core_enable(sdev, swidget);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: this function returns an int so presumably can fail? should the error handling be added here?

File: sound/soc/sof/topology.c [hunk 14]
```diff
@@ -1233,6 +1243,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
...
+{
+	struct sof_ipc_pm_core_config pm_core_config = {
+		.hdr = {
+			.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CORE_ENABLE,
+			.size = sizeof(pm_core_config),
+		},
+		.enable_mask = sdev->enabled_cores_mask | BIT(core),
+	};
+	int ret;
+
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1299]: error check if core isn't in the range of platform-supported cores? we have a cores_mask field for Intel.

File: sound/soc/sof/topology.c [hunk 15]
```diff
@@ -2159,7 +2203,7 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
  */
 
 static int sof_widget_load_process(struct snd_soc_component *scomp, int index,
-				   struct snd_sof_widget *swidget,
+				   struct snd_sof_widget *swidget, int core,
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2249]: could we have all these additions of a new parameter as a separate patch, with the paramter set to zero, and later in another patch we set it a a non-zero value? It'd reduce the size of all this.

File: sound/soc/sof/sof-audio.c (hunk 12 above)
Comment (lyakh) [reply sound/soc/sof/sof-audio.c:160]: @plbossart it wasn't, no. You're probably thinking about `snd_sof_find_swidget_id()` and `snd_sof_find_dai_pipe()`

File: sound/soc/sof/sof-audio.c (hunk 13 above)
Comment (lyakh) [reply sound/soc/sof/sof-audio.c]: yep, added.

File: sound/soc/sof/topology.c (hunk 14 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:1299]: this should be done, yes, but I think this should be done inside `.core_power_up()` which is called just above and is arch-dependent, whereas this function is not. E.g. such a check should be added to `hda_dsp_enable_core()` but that should be a separate patch, since this is also relevant now, this isn't something new, added by this PR.

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:2249]: done

File: sound/soc/sof/topology.c (hunk 14 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:1299]: ok

File: sound/soc/sof/sof-audio.h [hunk 16]
```diff
@@ -151,6 +152,8 @@ int snd_sof_complete_pipeline(struct device *dev,
 int sof_load_pipeline_ipc(struct device *dev,
 			  struct sof_ipc_pipe_new *pipeline,
 			  struct sof_ipc_comp_reply *r);
+int sof_pipeline_core_enable(struct snd_sof_dev *sdev,
+			     const struct snd_sof_widget *swidget);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h:156]: ok, it's the end of my day and my brain is fried, so maybe a stupid question but shouldn't we have a symmetric sof_pipeline_core_disable()? Or if it's handled automagically add a comment maybe?

Comment (lyakh) [reply sound/soc/sof/sof-audio.h:156]: @plbossart I don't think this is currently supported. DSP cores are enabled when pipelines, that should run on them are loaded, not when they are used. And we don't load or unload pipelines dynamically so far. So, as soon as a topology with pipelines, running on slave cores is loaded, those cores will be enabled and they'll stay enabled until the DSP is turned off.

Comment (plbossart) [reply sound/soc/sof/sof-audio.h:156]: oh wow, that's a really big issue then. we should really plan for cores to be turned off when there's nothing scheduled on them. 

@lgirdwood need you to chime in...

Comment (lyakh) [reply sound/soc/sof/sof-audio.h:156]: @plbossart possibly, but this is really out of the scope of this PR. Note, that this PR doesn't *add* this functionality, it just extends it to the component level.

Comment (plbossart) [reply sound/soc/sof/sof-audio.h:156]: The ask was to support multi-core usages, but not to leave the cores on at all times, even when unused. Maybe that's a change of scope or an undocumented capability, but do you disagree this is needed?

Comment (lyakh) [reply sound/soc/sof/sof-audio.h:156]: @plbossart I don't disagree with that. This PR isn't by any means a complete multi-core support. This is just the first step. After it we can decide which further steps we want to make and how. For this feature - powering down slave cores when not in use - I think we'll need something like or indeed per-core runtime PM. Powering up a core isn't as simple as powering up a simple fixed-function device. When powered up slave cores execute their iniitialisation code, allocate memory, interrupts, initiialise scheduling etc. Also the notion of a "core in use" might be not so simple. E.g. as far as I understand the current plan is to support for example pipelines, running on the master core, where just one or more components from it run on slave cores. So, each pipeline would have a bit-mask of used cores. All this is future development.

Comment (plbossart) [reply sound/soc/sof/sof-audio.h:156]: ok. do you mind adding comments on current limitations, both in code and commit messages, so that upstream reviewers are not surprised when the second batch comes their way?

@lgirdwood @mmaka1 @lbetlej we need to have a discussion on this, there can be multiple ways of dealing with core power management.

Comment (lyakh) [reply sound/soc/sof/sof-audio.h:156]: @plbossart done

**Comment by lyakh** [comment]
@lgirdwood this is the kernel PR that adds the new `SOF_TKN_COMP_CORE_ID` token

**Comment by kv2019i** [comment]
SO ready to merge. What is the story with CI failures on CML_HEL_RT5682 ? I see then on multiple test runs? @lyakh

**Comment by kv2019i** [comment]
SOFCI TEST

**Comment by lyakh** [comment]
> Partial agree to unlock firmware integration and tests. I don't think we should send this upstream until we have at least a story on the 'advanced DSP core power management'.

@plbossart I'd be happy to have such a story too, but most of this work is taking part in the firmware. E.g. @mrajwa knows more about that

**Comment by lyakh** [comment]
> SO ready to merge. What is the story with CI failures on CML_HEL_RT5682 ? I see then on multiple test runs? @lyakh

@kv2019i a DSP panic... nice... looking, maybe topology has been modified in an "old" way

**Comment by lyakh** [comment]
> > SO ready to merge. What is the story with CI failures on CML_HEL_RT5682 ? I see then on multiple test runs? @lyakh
> 
> @kv2019i a DSP panic... nice... looking, maybe topology has been modified in an "old" way

what is a CML_HEL_RT5682 / sh-cml-hel-rt5682-01? I don't see it tested in other PRs

**Comment by lyakh** [comment]
> > > SO ready to merge. What is the story with CI failures on CML_HEL_RT5682 ? I see then on multiple test runs? @lyakh
> > 
> > 
> > @kv2019i a DSP panic... nice... looking, maybe topology has been modified in an "old" way
> 
> what is a CML_HEL_RT5682 / sh-cml-hel-rt5682-01? I don't see it tested in other PRs

@kv2019i ok, that buggy board is gone. Looks better now

**Comment by kv2019i** [comment]
@ranj063 Feel free to merge, I won't do it now in case you are in middle of rebase.

**Comment by wenqingfu** [comment]
> > > > SO ready to merge. What is the story with CI failures on CML_HEL_RT5682 ? I see then on multiple test runs? @lyakh
> > > 
> > > 
> > > @kv2019i a DSP panic... nice... looking, maybe topology has been modified in an "old" way
> > 
> > 
> > what is a CML_HEL_RT5682 / sh-cml-hel-rt5682-01? I don't see it tested in other PRs
> 
> @kv2019i ok, that buggy board is gone. Looks better now

CML_HEL was removed from CI for two days while testing a new tool that can offline/online a DUT model with a mouse click.  This model is on the critical path of testing and stretched the testing duration by several minutes. Unfortunately, this removal caused confusion to this thread that the issue is gone...  We brought it back for testing purpose. The issue is still there according to #2042
